File: android-core/src/androidTest/java/com/mparticle/internal/MessageManagerTests.java
package com.mparticle.internal;

import android.content.Context;
import android.content.SharedPreferences;

import com.mparticle.AccessUtils;
import com.mparticle.testutils.BaseCleanInstallEachTest;
import com.mparticle.MParticle;
import com.mparticle.MParticleOptions;
import com.mparticle.internal.database.tables.mp.SessionTable;
import com.mparticle.internal.networking.BaseMPMessage;
import com.mparticle.testutils.MParticleUtils;
import com.mparticle.testutils.mock.MockContext;

import junit.framework.Assert;

import org.junit.Test;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;

public class MessageManagerTests extends BaseCleanInstallEachTest {

    @Test
    public void testInstallReferrerMigration() throws Exception {
        //test previously installed
        setFirstRunLegacy(true, "key");
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        assertTrue(getMessageManager().isFirstRunForMessage());
        assertTrue(getMessageManager().isFirstRunForAST());

        setFirstRunLegacy(false, "key");
        MParticle.setInstance(null);
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        assertFalse(getMessageManager().isFirstRunForMessage());
        assertFalse(getMessageManager().isFirstRunForAST());
    }

    @Test
    public void testInstallReferrerFlags() throws Exception {
        // test that both AST and First Run flags get properly flipped when their
        // corresponding setters get called
        setFirstRunLegacy(true, "key");
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        assertTrue(getMessageManager().isFirstRunForAST());
        getMessageManager().setFirstRunForAST(false);
        assertFalse(getMessageManager().isFirstRunForAST());

        assertTrue(getMessageManager().isFirstRunForMessage());
        getMessageManager().setFirstRunForMessage(false);
        assertFalse(getMessageManager().isFirstRunForMessage());

        // same thing, but make sure nothing funky happens when the order gets
        // reversed
        MParticle.setInstance(null);
        setFirstRunLegacy(true, "key");
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());

        assertTrue(getMessageManager().isFirstRunForMessage());
        getMessageManager().setFirstRunForMessage(false);
        assertFalse(getMessageManager().isFirstRunForMessage());

        assertTrue(getMessageManager().isFirstRunForAST());
        getMessageManager().setFirstRunForAST(false);
        assertFalse(getMessageManager().isFirstRunForAST());

        // test that the flags remain persisted when the application restarts
        MParticle.setInstance(null);
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());

        assertFalse(getMessageManager().isFirstRunForMessage());
        assertFalse(getMessageManager().isFirstRunForAST());

    }

    @Test
    public void testDuplicateSessionEnds() throws Exception {
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        Session session = new Session();
        session.start(new MockContext());
        getMessageManager().startSession(session);
        MParticleUtils.awaitStoreMessage();
        BaseMPMessage message = getMessageManager().getMParticleDBManager().getSessionForSessionEndMessage(session.mSessionID, null, session.getMpids());
        Assert.assertNotNull(message);
        getMessageManager().getMParticleDBManager().updateSessionStatus(session.mSessionID, SessionTable.SessionTableColumns.STATUS);
        message = getMessageManager().getMParticleDBManager().getSessionForSessionEndMessage(session.mSessionID, null , session.getMpids());
        Assert.assertNull(message);
    }

    @Test
    public void testInstallReferrerFlagsEdgeCases() throws Exception {
        setFirstRunLegacy(true, "key");

        //start SDK, and send only AST, but not First Run message
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        assertTrue(getMessageManager().isFirstRunForAST());
        getMessageManager().setFirstRunForAST(false);
        assertFalse(getMessageManager().isFirstRunForAST());

        //restart SDK, and assert that AST message has been flagged, but First Run message has not
        MParticle.setInstance(null);
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());

        assertTrue(getMessageManager().isFirstRunForMessage());
        assertFalse(getMessageManager().isFirstRunForAST());


        setFirstRunLegacy(true, "key");

        //start SDK, and send only First Run message, but not AST
        MParticle.setInstance(null);
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());
        assertTrue(getMessageManager().isFirstRunForMessage());
        getMessageManager().setFirstRunForMessage(false);
        assertFalse(getMessageManager().isFirstRunForMessage());

        //restart SDK, and assert that First Run message has been flagged as sent, but AST has not
        MParticle.setInstance(null);
        assertNull(MParticle.getInstance());
        MParticle.start(MParticleOptions.builder(mContext)
                .credentials("key", "secret")
                .build());

        assertFalse(getMessageManager().isFirstRunForMessage());
        assertTrue(getMessageManager().isFirstRunForAST());

    }

    /**
     * simulates the install state to settings pre 5.0.9 || pre 4.17.1
     * @param firstRun
     */
    private void setFirstRunLegacy(boolean firstRun, String key) {
        SharedPreferences sharedPreferences = mContext.getSharedPreferences(Constants.PREFS_FILE, Context.MODE_PRIVATE);
        sharedPreferences.edit()
                .remove(Constants.PrefKeys.FIRSTRUN_AST + key)
                .remove(Constants.PrefKeys.FIRSTRUN_MESSAGE + key)
                .apply();
        if (firstRun) {
            sharedPreferences.edit().remove(Constants.PrefKeys.FIRSTRUN_OBSELETE + key).apply();
        } else {
            sharedPreferences.edit().putBoolean(Constants.PrefKeys.FIRSTRUN_OBSELETE + key, false).apply();
        }
    }

    private MessageManager getMessageManager() {
        return AccessUtils.getMessageManager();
    }
}

File: android-core/src/androidTest/java/com/mparticle/internal/MessageManagerTests.kt
package com.mparticle.internal

import android.content.Context
import android.content.SharedPreferences
import com.mparticle.AccessUtils
import com.mparticle.MParticle
import com.mparticle.MParticleOptions
import com.mparticle.internal.database.tables.SessionTable
import com.mparticle.internal.messages.BaseMPMessage
import com.mparticle.testing.BaseTest
import com.mparticle.testing.context
import com.mparticle.utils.startMParticle
import org.junit.Assert
import org.junit.Test

class MessageManagerTests : BaseTest() {
    @Test
    @Throws(Exception::class)
    fun testInstallReferrerMigration() {
        val apiKey = "apiKey"
        val options = MParticleOptions.builder(context)
            .credentials(apiKey, "secret")
        // test previously installed
        setFirstRunLegacy(true, apiKey)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        messageManager.isFirstRunForAST = true
        Assert.assertTrue(messageManager.isFirstRunForMessage)
        Assert.assertTrue(messageManager.isFirstRunForAST)
        setFirstRunLegacy(false, apiKey)
        MParticle.setInstance(null)
        startMParticle(options)
        Assert.assertFalse(messageManager.isFirstRunForMessage)
        Assert.assertFalse(messageManager.isFirstRunForAST)
    }

    @Test
    @Throws(Exception::class)
    fun testInstallReferrerFlags() {
        // test that both AST and First Run flags get properly flipped when their
        // corresponding setters get called
        val apiKey = "apiKey"
        val options = MParticleOptions.builder(context)
            .credentials(apiKey, "secret")
        setFirstRunLegacy(true, apiKey)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        messageManager.isFirstRunForAST = true
        Assert.assertTrue(messageManager.isFirstRunForAST)
        messageManager.isFirstRunForAST = false
        Assert.assertFalse(messageManager.isFirstRunForAST)
        Assert.assertTrue(messageManager.isFirstRunForMessage)
        messageManager.isFirstRunForMessage = false
        Assert.assertFalse(messageManager.isFirstRunForMessage)

        // same thing, but make sure nothing funky happens when the order gets
        // reversed
        MParticle.setInstance(null)
        setFirstRunLegacy(true, "apiKey")
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        Assert.assertTrue(messageManager.isFirstRunForMessage)
        messageManager.isFirstRunForMessage = false
        Assert.assertFalse(messageManager.isFirstRunForMessage)
        Assert.assertTrue(messageManager.isFirstRunForAST)
        messageManager.isFirstRunForAST = false
        Assert.assertFalse(messageManager.isFirstRunForAST)

        // test that the flags remain persisted when the application restarts
        MParticle.setInstance(null)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        Assert.assertFalse(messageManager.isFirstRunForMessage)
        Assert.assertFalse(messageManager.isFirstRunForAST)
    }

    @Test
    @Throws(Exception::class)
    fun testDuplicateSessionEnds() {
        startMParticle()
        val session = InternalSession()
        session.start(context)
        messageManager.startSession(session)
        com.mparticle.internal.AccessUtils.awaitMessageHandler()
        var message: BaseMPMessage? =
            messageManager.mParticleDBManager.getSessionForSessionEndMessage(
                session.mSessionID,
                null,
                session.mpids
            )
        junit.framework.Assert.assertNotNull(message)
        messageManager.mParticleDBManager.updateSessionStatus(
            session.mSessionID,
            SessionTable.SessionTableColumns.STATUS
        )
        message = messageManager.mParticleDBManager.getSessionForSessionEndMessage(
            session.mSessionID,
            null,
            session.mpids
        )
        junit.framework.Assert.assertNull(message)
    }

    @Test
    @Throws(Exception::class)
    fun testInstallReferrerFlagsEdgeCases() {
        val apiKey = "apiKey"
        val options = MParticleOptions.builder(context)
            .credentials(apiKey, "secret")
        setFirstRunLegacy(true, "key")

        // start SDK, and send only AST, but not First Run message
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        messageManager.isFirstRunForAST = true
        Assert.assertTrue(messageManager.isFirstRunForAST)
        messageManager.isFirstRunForAST = false
        Assert.assertFalse(messageManager.isFirstRunForAST)

        // restart SDK, and assert that AST message has been flagged, but First Run message has not
        MParticle.setInstance(null)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        Assert.assertTrue(messageManager.isFirstRunForMessage)
        Assert.assertFalse(messageManager.isFirstRunForAST)
        setFirstRunLegacy(true, apiKey)

        // start SDK, and send only First Run message, but not AST
        MParticle.setInstance(null)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        Assert.assertTrue(messageManager.isFirstRunForMessage)
        messageManager.isFirstRunForMessage = false
        Assert.assertFalse(messageManager.isFirstRunForMessage)

        // restart SDK, and assert that First Run message has been flagged as sent, but AST has not
        MParticle.setInstance(null)
        Assert.assertNull(MParticle.getInstance())
        startMParticle(options)
        Assert.assertFalse(messageManager.isFirstRunForMessage)
        Assert.assertTrue(messageManager.isFirstRunForAST)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testDelayedStartInvoked() {
        val options = MParticleOptions.builder(context).credentials("key", "secret").build()
        MParticle.start(options)
        val messageManager = AccessUtils.messageManager
        Assert.assertFalse(messageManager.hasDelayedStartOccurred())
        val handler = com.mparticle.internal.AccessUtils.messageHandler
        handler.sendMessage(handler.obtainMessage())
        com.mparticle.internal.AccessUtils.awaitMessageHandler()
        Assert.assertTrue(messageManager.hasDelayedStartOccurred())
    }

    /**
     * simulates the install state to settings pre 5.0.9 || pre 4.17.1
     * @param firstRun
     */
    private fun setFirstRunLegacy(firstRun: Boolean, key: String) {
        val sharedPreferences: SharedPreferences =
            context.getSharedPreferences(Constants.PREFS_FILE, Context.MODE_PRIVATE)
        sharedPreferences.edit()
            .remove(Constants.PrefKeys.FIRSTRUN_AST + key)
            .remove(Constants.PrefKeys.FIRSTRUN_MESSAGE + key)
            .apply()
        if (firstRun) {
            sharedPreferences.edit().remove(Constants.PrefKeys.FIRSTRUN_OBSELETE + key).apply()
        } else {
            sharedPreferences.edit().putBoolean(Constants.PrefKeys.FIRSTRUN_OBSELETE + key, false)
                .apply()
        }
    }

    private val messageManager: MessageManager
        private get() = AccessUtils.messageManager
}
